diff --git a/__init__.py b/usr/local/lib/python3.12/dist-packages/torch/accelerator/__init__.py
index 8f912b4..b8aa0d9 100644
--- a/__init__.py
+++ b/usr/local/lib/python3.12/dist-packages/torch/accelerator/__init__.py
@@ -43,7 +43,7 @@ def is_available() -> bool:
 
         >>> assert torch.accelerator.is_available() "No available accelerators detected."
     """
-    return device_count() > 0
+    return torch.cuda.is_available()
 
 
 def current_accelerator() -> torch.device:
