diff --git a/lib/driver/llvm.cc b/lib/driver/llvm.cc
index a73e6541d..01e38f825 100644
--- a/lib/driver/llvm.cc
+++ b/lib/driver/llvm.cc
@@ -25,6 +25,7 @@
 #endif
 #include <memory>
 #include <regex>
+#include <iostream>
 #include "triton/driver/llvm.h"
 #include "triton/driver/dispatch.h"
 #include "triton/driver/error.h"
@@ -148,7 +149,9 @@ namespace triton
 
     int vptx(int version)
     {
-      if (version >= 11040)
+      if (version >= 12000)
+        return 80;
+      else if (version >= 11040)
         return 74;
       // if(version >= 11030) return 73;
       // if(version >= 11020) return 72;
@@ -163,15 +166,15 @@ namespace triton
     std::string llir_to_ptx(llvm::Module *module, int cc, int version)
     {
       // LLVM version in use may not officially support target hardware
-      int max_nvvm_cc = 75;
-      int max_nvvm_ptx = 74;
+      int max_nvvm_cc = 90;
+      int max_nvvm_ptx = 80;
       // options
       auto options = llvm::cl::getRegisteredOptions();
       auto *short_ptr = static_cast<llvm::cl::opt<bool> *>(options["nvptx-short-ptr"]);
       assert(short_ptr);
       short_ptr->setValue(true);
       // compute capability
-      std::string sm = "sm_" + std::to_string(cc);
+      std::string sm = cc == 90 ? "sm_90a" : "sm_" + std::to_string(cc);
       // max PTX version
       int ptx = vptx(version);
       int ptx_major = ptx / 10;
@@ -244,7 +247,9 @@ namespace triton
       ofs.close();
       std::string cmd;
       int err;
-      cmd = ptxas + " -v --gpu-name=sm_" + std::to_string(cc) + " " + fsrc + " -o " + fsrc + ".o 2> " + flog;
+      cmd = cc == 90 ? 
+        ptxas + " -v --gpu-name=sm_90a " + fsrc + " -o " + fsrc + ".o 2> " + flog :
+        ptxas + " -v --gpu-name=sm_" + std::to_string(cc) + " " + fsrc + " -o " + fsrc + ".o 2> " + flog;
       err = system(cmd.c_str());
       if (err != 0)
       {
